### ISSUE TITLE
Fonts list, and suggestions

### ISSUE THREAD

**Issue by ghost** [NiklasGollenstede/stop-fingerprinting#2]
Thanks for this useful addon.

I run it through the tests at https://browserprint.info and wanted to share a few ideas in case you're accepting requests:
- let the user specify a "fixed" fonts list for the field "Fonts CSS only" in browserprint (in my case, there are so many fonts installed that even randomizing the list doesn't seem to reduce the unique fingerprint)
- block or randomize the "Character Sizes" field, if possible
- block the HTML5 Audio API to stop the "Audio Fingerprints" (I don't know if this can be done; I've tried asking here for now: https://bugzilla.mozilla.org/show_bug.cgi?id=1288359)

And a few extra settings you may want to disable (used for fingerprinting when active):
dom.battery.enabled
device.sensors.enabled
camera.control.face_detection.enabled
dom.enable_resource_timing
dom.enable_user_timing

**Comment by NiklasGollenstede** [comment]
I am Indeed interested in feedback, so thanks for that.


> let the user specify a "fixed" fonts list for the field "Fonts CSS only" in browserprint (in my case, there are so many fonts installed that even randomizing the list doesn't seem to reduce the unique fingerprint)


You want that your browser only uses a subset of the available fonts at all, have a look at browser.display.use_document_fonts (you can't choose the list, though)

> block or randomize the "Character Sizes" field, if possible


I have no idea what you mean

> block the HTML5 Audio API


I have this on my agenda, but expect it to be rather  complicated.


> And a few extra settings

I've changed over to the WebExtension API, which can't change config entries. But I thimk I will find other ways to disable/fake these.
(For example: disabling the battery API will make a desktop computer even more unique. It should rather be "reset" to some standard or randomized)

**Comment by bryce-lynch** [comment]
> > block or randomize the "Character Sizes" field, if possible
> I have no idea what you mean

I mean the "Character Sizes" listed on browserprint's results page. It's similar to the fonts list, but it lists the sizes of available Unicode character (e.g. 424x800). I don't know how it is used for fingerprinting, but I assume it varies from system to system.

> disabling the battery API will make a desktop computer even more unique

But since all this canvas/audio/battery fingerprinting stuff appeared with html5 APIs, I guess pre-html5 browsers would have simply reported them as disabled. Although I don't know if there exist web engines with no html5 support...

Anyway, thanks a lot for your work.

**Comment by NiklasGollenstede** [comment]
> browserprint's results

Can you tell me how exactly I get results on that page? On https://browserprint.info/view it just keeps telling me I need to have cookies enabled, even though I do. I can't find a better link.

**Comment by bryce-lynch** [comment]
Try from here: https://browserprint.info/captcha

**Comment by NiklasGollenstede** [comment]
Ah, thanks.
Regarding the "Character Sizes" thing: I think it's quite strictly related to the font detection. The randomization that is implemented in the version of this add-on that is on AMO should already defeat that, doesn't it?

And yes, all major browsers support (most) HTML5, so removing any of those APIs definitely makes you stand out.

**Comment by bryce-lynch** [comment]
I'm not totally sure about the randomization. When the addon is disabled, I only see 6 fonts listed in "Fonts (CSS only)". When the addon is enabled, "Fonts (CSS only)" becomes a huge (5 KB long) list, while "Character Sizes" remains identical to before.

**Comment by NiklasGollenstede** [comment]
> huge (5 KB long) list

Hmm. The current implementation gives each not installed font a chance of being detected as installed. The height of that chance depends on the exact detection mechanism. Apparently browserprint.info tests a lot of fonts, and detects ether all or most of them as installed. Ether way, your truly installed fonts should be well hidden.

> "Character Sizes" remains identical to before

I'll look into that in the next days

**Comment by bryce-lynch** [comment]
> I'll look into that in the next days

Thanks. Anyway it seems Character Sizes is dependent on zoom, and is explained here https://www.bamsoftware.com/talks/fc15-fontfp/fontfp.html (there is also browserprint's source code on github if you're interested).

**Comment by bryce-lynch** [comment]
FYI I was able to stop audiocontext() fingerprinting using this very simple Greasemonkey user script:
Object.defineProperty(window, "AudioContext", {value: []});

However, DinamicsCompressor tests are still working as they don't seem to depend on audiocontext.

**Comment by NiklasGollenstede** [comment]
Yes, disabling an API by removing the reference to is quite easy. However, you have to make sure that:
1. you remove ALL references to it
2. do so before any page script is able to run and grab a reference
3. it's done for every frame (also see 2.)

If you miss any of those, it requires very little effort from the tracker to circumvent your "blockade".

And besides, blocking the was you did it will obviously break all usage of the API, even if it only uses parts that can't be used maliciously.

But yes. It will probably do for now.

> However, DinamicsCompressor tests are still working as they don't seem to depend on audiocontext.

To stop this you have to find out which function on which ``.prototype`` of which audio node or whatever is mandatory to the fingerprinting and block that.
E.g. ``Object.defineProperty(Audio<Whatever>.prototype, "<someFinction>", { value: function() { return something || null; }, });``

**Comment by bryce-lynch** [comment]
> If you miss any of those, it requires very little effort from the tracker to circumvent your "blockade".
Unfortunately that's beyond my capabilities. I'm no developer, just put together a quick & dirty hack (I don't care if those APIs break as I don't need them and I always keep things like webgl, webrtc, etc. turned off), while I wait for more skilled people to implement real anti-fingerprinting for modern web browsers.

**Comment by adeorato** [comment]
After reading your discussion I used now https://panopticlick.eff.org to compare the system font.

The result for system font were (one in x browsers have this value, so smaller is better):
FF without stop-fingerprinting 36167.6
FF with  stop-fingerprinting:  180812.0 
TOR browser bundle: 65.23

So just looking on this page and system font, this add-on increases the fingerprint by rounded 5 times compared to none anti fingerprinting add-on. 

My suggestion is to use the system font list of either
a) a fresh installed windows (separate list for each user-agent, maintainable?)
b) The list of TOR (my preferred way)
c) don't change it and leave the original values

Whatever you choose it may be a good idea to recheck the result against the above mentioned page.
I hope you consider my feedback and thank you for taking your time to fight against fingerprinting.

**Comment by NiklasGollenstede** [comment]
Unfortunately I still haven't found a real solution to the font detection problem.
It would be best if one could replace the list of "installed" fonts, indeed. Unfortunately I think that the Firefox code that finds the locally installed fonts is purely c++, and thus inaccessible for add-ons. The only solution would be to make and maintain custom builds of Firefox ... 
I am looking into "hackish" ways to emulate that list, but so far I've always been able to outsmart all of my own ideas -.-

Anyway. **If** there is a way to replace the font list, there is still the question of what to replace it with.
I think a very important base assumption to make here is: de-facto it is not possible to get rid of *all* fingerprinting vectors (there is simply too much stuff that can potentially be used).
Not changing the fonts certainly isn't a solution, it won't reduce fingerprintability at all.
Using the TOR list certainly is one approach, but the problem is that it will simply put you from the pool of (e.g.) "Windows 7 Firefox users that have OpenOffice installed" into the pool of "TOR-Browser users". And don't be fooled by the numbers, in practice the latter pool isn't necessarily larger. The statistics on all the fingerprinting analysis pages are heavyly biased towards users that do use measures against fingerprinting. Furthermore it is impossible to be sure that you didn't miss some minor aspect which your test don't test either but which sets you apart from others because the trackers do look at that aspect.
Which leaves the approach to create a random list (for every session/window/tab/page load/...), which I still think is the best solution. Any "real" information will be very hard to single out in all the random noise, so I think this method is much more forgiving towards mistakes made by the anti-fingerprinting software (any every software contains bugs). And a "one in a gazillion browsers have the same print" isn't a problem because the next time you will have a different fingerprint in the gazillion.
If faking a fixed list is wearing latex gloves, then randomization is wearing Mission Impossible style glued-on fingerprints, which change into a new fingerprint every time it matters. If someone analyses a crime scene without finding any fingerprints, they will look for boot prints and DNA and what not. But on the other hand if they find authentic fingerprints, they are much more likely to be satisfied with that for the moment and base their further investigation on that.

> So just looking on this page and system font, this add-on increases the fingerprint by rounded 5 times compared to none anti fingerprinting add-on. 

Yes, every individual fingerprint you get is much more recognizable than by default. But it isn't worth anything because the next time you'll get a new one.

**Comment by Atavic** [comment]
Character Sizes: [DEPRECATED] The height and width of a set of Unicode characters when rendered with a set of different styles (e.g. sans-serif). Deprecated because in the current implementation zooming changes the result.
